samtranslator/plugins/api/implicit_api_plugin.py

```diff
@@ -173,8 +173,19 @@ def _add_api_to_swagger(self, event_id, event_properties, template):
         # Need to grab the AWS::Serverless::Api resource for this API event and update its Swagger definition
         api_id = self._get_api_id(event_properties)
 
+        # As of right now, this is for backwards compatability. SAM fails if you have an event type "Api" but that
+        # references "AWS::Serverless::HttpApi". If you do the opposite, SAM still outputs a valid template. Example of that
+        # can be found https://github.com/aws/serverless-application-model/blob/develop/tests/translator/output/api_with_any_method_in_swagger.json.
+        # One would argue that, this is unexpected and should actually fail. Instead of suddenly breaking customers in this
+        # position, we added a check to make sure the Plugin run (Http or Rest) is referencing an api of the same type.
```

Comment:
How is the HttpApi event referencing an Api scenario being addressed? As far as I understand, this condition only checks for Api->HttpApi.

Reply:
Through the condition: 

```
is_referencing_http_from_api_event = (
            not template.get(api_id)
            or template.get(api_id).type == "AWS::Serverless::HttpApi"
            and not template.get(api_id).type == self.api_type
        )
```

So this `_add_api_to_swagger` is called when the `AWS::Serverless::Function.Event.Type` is `Api` or `HttpApi`, so we can generate the OpenApi (old name swagger). We are checking to see if the referenced Api is or type `AWS::Serverless::HttpApi` and the current API Plugin type matches the one it references.